File: frontend/lib/persona.ts
/** Persona defaults - the frontend half of `backend/src/cinqflow/auth/persona.py`.
 *
 *  The mapping from roles to persona and capabilities lives in exactly one
 *  place, the backend; `CurrentUser` arrives with both already decided. This
 *  module only turns a persona into *defaults* - which reading mode a review
 *  opens in, which proposal filter, whether technical columns start collapsed,
 *  whether the workflow's unreached steps are shown - and holds the two
 *  sentences the UI says when a control isn't the caller's.
 *
 *  Persona is emphasis; capability is authority. Nothing here hides a fact or
 *  grants anything: the API enforces `capabilities`, the UI mirrors them
 *  (plan §14.2, §18.2). The analyst's own reading-mode override
 *  (`localStorage`, see `ReadingMode`) still wins over the default. */

import type { ReadingModeKey } from "@/components/run/ReadingMode";
import type { Persona } from "@/lib/auth";

export const PERSONA_LABEL: Record<Persona, string> = {
  data_analyst: "Data Analyst",
  data_platform: "Data Platform",
};

export interface PersonaDefaults {
  readingMode: ReadingModeKey;
  /** S4: start on the fields that need a decision, or on all of them. */
  proposalFilter: "decisions" | "all";
  /** Forensic tables: technical/system columns start collapsed (lands with PR-7). */
  technicalCollapsed: boolean;
  /** `WorkflowSteps`: show every declared step, including the unreached ones,
   *  from the start (Data Platform) - or only what has happened (Data Analyst). */
  workflowStepsExpanded: boolean;
}

export function personaDefaults(persona: Persona): PersonaDefaults {
  switch (persona) {
    case "data_platform":
      return {
        readingMode: "forensic",
        proposalFilter: "all",
        technicalCollapsed: false,
        workflowStepsExpanded: true,
      };
    case "data_analyst":
    default:
      return {
        readingMode: "evidence",
        proposalFilter: "decisions",
        technicalCollapsed: true,
        workflowStepsExpanded: false,
      };
  }
}

/** Shown in place of a gate's controls when `can_decide_gates` is false. */
export const GATE_LOCKED_REASON =
  "Your role can review this run but not decide it — an approver or business analyst signs the gate.";

/** Shown in place of a retry/re-run control when `can_rerun_steps` is false. */
export const RERUN_LOCKED_REASON =
  "Retrying and re-running are Data Platform actions — a data engineer or operations role.";

File: backend/src/cinqflow/auth/persona.py
"""Persona and capabilities, derived from roles - the one mapping, owned here.

Two ideas, kept deliberately apart (docs/blueprints/Analyst_worflow_and_DAGs/
04_validation_and_implementation_plan.md §14):

- **Persona is emphasis.** It picks defaults - which reading mode a review opens
  in, which proposal filter, which home page. Two personas: the Data Analyst
  (onboards a file and signs the gates) and the Data Platform (keeps the
  workflow running and re-runs what failed). Strictly role-derived, no switcher
  (decision D2).
- **Capability is authority.** Whether a caller may *decide a gate* or *re-run a
  step* is a role predicate the API enforces (`require_capability`), independent
  of persona. An administrator who also holds `approver` can approve; a
  `data_steward` can review but not decide - and the screen says so.

Precedence for a user holding roles from both sides: any platform role wins. A
user with no roles at all is a Data Analyst with no capabilities - view only.
"""

from __future__ import annotations

from collections.abc import Iterable
from typing import Literal

from pydantic import BaseModel

Persona = Literal["data_analyst", "data_platform"]

#: Decision D1. Everything not listed here is an analyst role (or unknown, which
#: also reads as analyst with no capabilities - the safe default).
PLATFORM_ROLES: frozenset[str] = frozenset({"data_engineer", "operations", "administrator"})
ANALYST_ROLES: frozenset[str] = frozenset(
    {"business_analyst", "approver", "data_steward", "read_only"}
)

#: May decide G1 and G2. `data_steward` owns the knowledge plane, not the gate;
#: `administrator` runs the system, not the data - neither signs a gate unless it
#: also holds one of these (forward-flow-adoption.md §7.1).
GATE_ROLES: frozenset[str] = frozenset({"business_analyst", "approver"})

#: May retry / re-run a workflow step (`/retry` today; `/steps/{step}/rerun` from
#: PR-3). Re-running is an operational act on the platform, hence the platform set.
RERUN_ROLES: frozenset[str] = PLATFORM_ROLES


class Capabilities(BaseModel):
    """What this caller may *do*. Rendered by the UI as affordances, enforced by
    the API as 403s - the UI mirrors, the API decides."""

    can_decide_gates: bool
    can_rerun_steps: bool
    can_manage_users: bool


def persona_for(roles: Iterable[str]) -> Persona:
    held = set(roles)
    return "data_platform" if held & PLATFORM_ROLES else "data_analyst"


def capabilities_for(roles: Iterable[str]) -> Capabilities:
    held = set(roles)
    return Capabilities(
        can_decide_gates=bool(held & GATE_ROLES),
        can_rerun_steps=bool(held & RERUN_ROLES),
        can_manage_users="administrator" in held,
    )
